[SMTChecker] Use nonlinear clauses instead of inlining base constructors in CHC

leonardoalt opened the pull request:
CHC currently inlines all base constructors, which won't work for `return` edges.
In preparation for that, this PR replaces the CHC encoding of base constructors to nonlinear clauses "calling" the bases, which might be less efficient but is a lot more elegant in the encoding and well, makes a lot more sense in terms of control flow anyway.
This is fundamentally needed so we can implement `return` in CHC.

This PR still needs a little cleanup, but the core can already be reviewed.

Commit:
Use nonlinear clauses instead of inlining for base constructors

mijovic commented on libsmtutil/Z3CHCInterface.cpp:
```diff
@@ -58,6 +58,7 @@ void Z3CHCInterface::registerRelation(Expression const& _expr)
 void Z3CHCInterface::addRule(Expression const& _expr, string const& _name)
 {
 	z3::expr rule = m_z3Interface->toZ3Expr(_expr);
+	//cout << rule << "\n\n";
```
I guess this will be removed

mijovic commented on libsolidity/formal/PredicateInstance.cpp:
```diff
@@ -128,9 +164,29 @@ vector<smtutil::Expression> currentFunctionVariables(
+)
+{
+	auto& state = _context.state();
+	vector<smtutil::Expression> exprs{_context.state().errorFlag().currentValue(), state.thisAddress(0), state.crypto(0), state.tx(0), state.state()};
```
Maybe split into more lines, plus use state in first argument as well.
```suggestion
	vector<smtutil::Expression> exprs{
		state.errorFlag().currentValue(), 
		state.thisAddress(0), 
		state.crypto(0), 
		state.tx(0), 
		state.state()
	};
```

Maybe one more idea is to create a method in state so that it returns list of expressions for this case.

leonardoalt replied:
The annoying part is it's not always the same subset that we're interested in. For example, here for the last one we want `state.state()`, but in the next function we want `state.state(0)`

mijovic replied:
I see :)

mijovic commented on libsolidity/formal/CHC.cpp:
```diff
@@ -112,72 +112,182 @@ vector<string> CHC::unhandledQueries() const
-		implicitConstructorSort(state()),
-		"implicit_constructor_" + contractSuffix(_contract),
-		PredicateType::ImplicitConstructor,
+	if (auto constructor = _contract.constructor())
```
```suggestion
	if (auto const* constructor = _contract.constructor())
```

mijovic replied:
Not so important, also it is used without `const*` in the rest of the file, so I guess we leave it as is

leonardoalt replied:
Yea, I think I'll leave without

mijovic commented on libsolidity/formal/CHC.cpp:
```diff
@@ -112,72 +112,182 @@ vector<string> CHC::unhandledQueries() const
+	setCurrentBlock(entry);
+
+	auto baseArgs = baseArguments(_contract);
+	//cout << "Checking base args for contract " << _contract.name() << endl;
+	//for (auto const& [contract, args]: baseArgs)
+	//	cout << "Contract " << contract->name() << " has " << args.size() << " args" << endl;
```
Just for reference, this should be also removed

mijovic commented on libsolidity/formal/PredicateInstance.cpp:
```diff
@@ -128,9 +164,29 @@ vector<smtutil::Expression> currentFunctionVariables(
+	exprs += applyMap(_function.parameters(), [&](auto _var) { return _context.variable(*_var)->currentValue(); });
+
+	state.newState();
+	
```
Coding style complains here I think
```suggestion

```

mijovic commented on libsolidity/formal/CHC.cpp:
```diff
@@ -112,72 +112,182 @@ vector<string> CHC::unhandledQueries() const
-		solAssert(_function.scope() != m_currentContract, "");
-		SMTEncoder::visit(_function);
+		/// Collect base contracts and potential constructor arguments.
+		for (auto inh: contract->baseContracts())
```
Can you change the name `inh` here to something more readable?

leonardoalt replied:
Changed

mijovic commented on libsolidity/formal/CHC.cpp:
```diff
@@ -112,72 +112,182 @@ vector<string> CHC::unhandledQueries() const
+		/// Collect base contracts and potential constructor arguments.
+		for (auto inh: contract->baseContracts())
+		{
+			solAssert(inh, "");
```
Can this happen at all?

mijovic replied:
I am not suggesting to remove assert, it should be there, just thinking when and how `null` would be added to the list :)

leonardoalt replied:
Yea I also don't know when it happens, but the function returns a pointer, so better test it before using it.

mijovic commented on libsolidity/formal/CHC.cpp:
```diff
@@ -112,72 +112,182 @@ vector<string> CHC::unhandledQueries() const
+		for (auto inh: contract->baseContracts())
+		{
+			solAssert(inh, "");
+			auto const& base = dynamic_cast<ContractDefinition const&>(*inh->name().annotation().referencedDeclaration);
```
maybe `baseContract`?

leonardoalt replied:
If that's not too important I'd prefer to leave it as `base`. The function is already related to inheritance and we're calling `baseContracts()`, so to me it's clear that it means base contract, but with the shorter variable name `base`.

mijovic commented:
Generally looks good.
Will need to do one more review round, as I used this one to understand the changes and to check some small thinks and style.

mijovic commented:
Also, I am a bit worried that one test fails only on OSX in `soltest`

leonardoalt commented:
The usual nondeterminism I guess...cli also fails with a crash, need to find out what file it is

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `6340d1eae64fe265d383024de32bef626b394f10`:
```
Error: Trailing whitespace found:
 libsolidity/formal/PredicateInstance.cpp:170: 

```
Please check that your changes are working as intended.

stackenbotten commented on libsolidity/formal/PredicateInstance.cpp:
```diff
@@ -128,9 +155,29 @@ vector<smtutil::Expression> currentFunctionVariables(
+	exprs += applyMap(_function.parameters(), [&](auto _var) { return _context.variable(*_var)->currentValue(); });
+
+	state.newState();
+	
```
Coding style error

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `f66e25345a05916764584dde45b72e9ee7b67f97`:
```
Coding style error:
 libsolidity/formal/CHC.h:34:#include "libsmtutil/SolverInterface.h"

```
Please check that your changes are working as intended.

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `901aab8ea2d5fb465d899501f0f275a10ea65fa8`:
```
Coding style error:
 libsolidity/formal/CHC.h:34:#include "libsmtutil/SolverInterface.h"

```
Please check that your changes are working as intended.

leonardoalt commented:
@mijovic fixed some of your comments, plus also did some refactoring:
- removed everything related to `implicit constructor`, since that's not used anymore
- extracted some constraints into the `initialConstraints` function
- moved the `baseArguments` function into `SMTEncoder`, maybe that can be reused by SMTEncoder itself or BMC

mijovic commented:
I have one idea for test to add. It shows difference between old codegen and one using ir.
```
contract A {
    uint x;
    constructor() {
        x = 42;
    }
    function f() public returns(uint256) {
        return x;
    }
}
contract B is A {
    uint public y = f();
}
contract C {
    B b = new B();
    function y() public {
        assert(b.y() == 42); // this should fail using oldcodegen and pass with ir
    }
}
```

I know right now SMTChecker should care about old codegen only, but eventually we will use ir by default. This test will be one of the things that would change.
I suggest that we sort these tests somewhere, or write in for of issue in github, or something...

Original test where you can see it is passing via IR is in `constructor_inheritance_init_order.sol`

leonardoalt commented on libsolidity/formal/CHC.cpp:
```diff
@@ -112,48 +112,65 @@ vector<string> CHC::unhandledQueries() const
+					{
+						args.at(i)->accept(*this);
+						solAssert(m_context.knownVariable(*params.at(i)), "");
+						m_context.addAssertion(currentValue(*params.at(i)) == expr(*args.at(i)));
```
Note to self: use `expr(*args.at(i), targetType)`

leonardoalt commented on libsolidity/formal/CHC.cpp:
```diff
@@ -890,6 +869,42 @@ void CHC::defineInterfacesAndSummaries(SourceUnit const& _source)
+		for (auto var: constructor->parameters())
+			preConstraints = preConstraints && m_context.variable(*var)->valueAtIndex(0) == currentValue(*var);
+	preConstraints = preConstraints && errorFlag().currentValue() == 0;
+	m_context.addAssertion(preConstraints);
```
Note to self: this block can use `initialConstraints(_contract)`

leonardoalt commented:
@mijovic I changed your test to
```

leonardoalt commented:
Added test and fixed review comments

mijovic commented on libsolidity/formal/CHC.cpp:
```diff
@@ -967,6 +986,25 @@ void CHC::connectBlocks(smtutil::Expression const& _from, smtutil::Expression co
+
+	FunctionDefinition const* function = _function ?
+		_function :
+		_contract.constructor() ?
+			_contract.constructor() :
+			nullptr;
```
I think it is enough to just set it to `_contract.constructor()` if `_function` is `nullptr`
```suggestion
		_contract.constructor();
```

mijovic commented on test/libsolidity/smtCheckerTests/inheritance/constructor_state_variable_init_chain_tree.sol:
```diff
@@ -0,0 +1,43 @@
+contract C is B {
+	constructor(int c) B(-c) {
+		if (x > 0) {
+			assert(c < 0); // should hold
```
It holds here and fails on next line?

leonardoalt replied:
Yes, should be. Anything wrong?

mijovic replied:
No, I was thinking that because of comment you wrote "should fail" that it maybe doesn't fail. All fine :)

mijovic commented on libsolidity/formal/CHC.cpp:
```diff
@@ -112,48 +112,84 @@ vector<string> CHC::unhandledQueries() const
+		// The loop below re-adds the constructor parameter=argument equality chain for all base contracts.
+		// This adds potentially useless constraints, but makes it possible not to have to recreate full
+		// inheritance chains for every contract.
+		for (auto base: _contract.annotation().linearizedBaseContracts)
```
You have already used `base` in upper loop, this made some bugs for sure

mijovic replied:
Anyway, would be nice to change it :)

leonardoalt commented:
@mijovic updated with your comments + finally fixing the bugs. Now it implements what we talked on Gitter:
- initialize all state variables as 0
- evaluate the base call arguments in the order (most derived) -> (most base)
- construct each contract in the order (most base) -> (most derived)

I also added more tests containing tests for the behavior above, including arguments with side-effects.

mijovic reviewed (dismissed):
Great work.
Checked all the tests, looks good. Implementation looks good.

Also, I think this doesn't need `Changelog` entry, so we can merge it now.

P.S. We need to cover same tests for codegen in semanticTests as well :)

leonardoalt commented on libsolidity/formal/CHC.cpp:
```diff
@@ -156,19 +158,21 @@ void CHC::endVisit(ContractDefinition const& _contract)
 				solAssert(params.size() == args.size(), "");
 				for (unsigned i = 0; i < params.size(); ++i)
+				{
+					args.at(i)->accept(*this);
```
@mijovic this is the fix for unnamed constructor parameters: the argument must be evaluated anyway, because of side effects.

leonardoalt commented on test/libsolidity/smtCheckerTests/inheritance/constructor_hierarchy_base_calls_with_side_effects_8.sol:
```diff
@@ -0,0 +1,22 @@
+
+contract A {
+	uint public x;
+	constructor(uint) {}
```
@mijovic in this case, before this commit, because the parameter is unnamed the argument `f()` was not evaluated, causing `x = 0` in the end which was wrong. Now it works.

leonardoalt commented on libsolidity/formal/CHC.cpp:
```diff
@@ -884,9 +893,15 @@ void CHC::defineContractInitializer(ContractDefinition const& _contract)
+		{
+			var->value()->accept(*this);
+			assignment(*var, *var->value());
+		}
```
@mijovic This is the fix for the variable initialization order. All state variables are already initialized as 0 before the whole inheritance deployment starts, which is correct. The wrong part here is that `SMTEncoder::initializeStateVariables` (removed line) first zero initializes then assigns values if available at state variable declaration. This is wrong here, since the zero init could overwrite other previous assignments which are valid. The correct behavior here is to only assign the declaration values, if available. In that case it is correct to overwrite previous assignments.

leonardoalt commented on libsolidity/formal/CHC.cpp:
```diff
@@ -884,9 +893,15 @@ void CHC::defineContractInitializer(ContractDefinition const& _contract)
 	setCurrentBlock(implicitConstructorPredicate);
 
+	solAssert(!m_errorDest, "");
+	m_errorDest = m_contractInitializers.at(&_contract);
```
@mijovic another bug was when there was a failing assertion inside an internal function called at state variable declaration, which is visited in the loop below. In that case, the error condition should jump only to the end of the contract initialization (routine defined in this function), and not to the constructor summary which represents the whole deployment procedure. Since this error destination is used by function call visits, this needs to be a member of the class. In fact this made the error destination actually simpler where it's used. The important part is to assign it properly for contract initialization, inheritance order, and function visit.

mijovic reviewed (commented):
I checked fixes only now, looks good. Just squashing and we should be fine finally

leonardoalt commented:
Squashed

mijovic commented:
Ah, `UNREACHABLE CODE WAS REACHED` again...

leonardoalt commented:
Yea I'm running it locally to check which test it is

leonardoalt commented:
Disabled test, let's see

ekpyron commented on libsolidity/formal/CHC.cpp:
```diff
@@ -112,48 +112,76 @@ vector<string> CHC::unhandledQueries() const
+		if (base != &_contract)
+		{
+			m_callGraph[&_contract].insert(base);
+			vector<ASTPointer<Expression>> const& args = baseArgs.count(base) ? baseArgs.at(base) : decltype(args){};
```
This might be https://github.com/ethereum/solidity/issues/10532

leonardoalt replied:
Still clang's fault though :p
It's all local so the ref should be fine

ekpyron replied:
https://en.cppreference.com/w/cpp/language/lifetime says "The lifetime of a temporary object may be extended by binding to a const lvalue reference or to an rvalue reference (since C++11), see reference initialization for details. "
It's an rvalue here, so it's new in C++11, maybe clang6 didn't quite catch that yet... although C++11 has been around for a while...

ekpyron replied:
By the way this will still always be a copy anyhow, though, I think.

ekpyron replied:
And I just confirmed in @agroce's docker image: removing the ref on the LHS fixes it...

leonardoalt replied:
Why would it be a copy in the true case?

ekpyron replied:
For the record: short version of the gitter explanation: the common type of the ternary is ``vector<...>`` (no ref), so the true-case constructs that from the const ref by invoking the copy constructor.
